src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -113,41 +120,38 @@ private void SetEntry(CacheEntry entry)
             // Initialize the last access timestamp at the time the entry is added
             entry.LastAccessed = utcNow;
 
-            var added = false;
             CacheEntry priorEntry;
-
-            _entryLock.EnterWriteLock();
-            try
+            if (_entries.TryGetValue(entry.Key, out priorEntry))
             {
+                priorEntry.SetExpired(EvictionReason.Replaced);
+            }
 
-                if (_entries.TryGetValue(entry.Key, out priorEntry))
+            if (!entry.CheckExpired(utcNow))
+            {
+                if (priorEntry == null ? _entries.TryAdd(entry.Key, entry) : _entries.TryUpdate(entry.Key, entry, priorEntry))
                 {
-                    _entries.Remove(entry.Key);
-                    priorEntry.SetExpired(EvictionReason.Replaced);
+                    entry.AttachTokens();
                 }
-
-                if (!entry.CheckExpired(utcNow))
+                else
                 {
-                    _entries[entry.Key] = entry;
-                    entry.AttachTokens();
-                    added = true;
+                    entry.SetExpired(EvictionReason.Replaced);
+                    entry.InvokeEvictionCallbacks();
                 }
             }
-            finally
+            else
             {
-                _entryLock.ExitWriteLock();
+                if (priorEntry != null)
+                {
+                    RemoveEntry(priorEntry);
```

Comment:
Yes but it's safe to call `InvokeEvictionCallbacks()` more than once. It should just no-op. Alternatively, `EntriesCollection.Remove(...)` can be used here instead